src/Voron/GlobalFlushingBehavior.cs

```diff
@@ -109,32 +108,32 @@ private void VoronEnvironmentFlushing()
             // ReSharper disable once FunctionNeverReturns
         }
 
-        private void SyncDesiredEnvironments(HashSet<StorageEnvironment> avoidDuplicates)
+        private void SyncEnvironments(bool force)
         {
-            avoidDuplicates.Clear();
-            EnvSyncReq envToSync;
-            var limit = _maybeNeedToSync.Count;
-            while (
-                // if there is high traffic into the queue, we want to abort after 
-                // we processed whatever was already in there, to avoid holding up
-                // the rest of the operations
-                limit-- > 0 &&
-                _maybeNeedToSync.TryDequeue(out envToSync))
+            foreach (var envSyncReq in _envsToSync)
             {
-                var storageEnvironment = envToSync.Env;
-                if (storageEnvironment == null)
+                var env = envSyncReq.Key.Owner;
+                if (env == null)
+                {
+                    _envsToSync.TryRemove(envSyncReq.Key, out _);
+                    continue;
+                }
+
+                if (env.Disposed)
                     continue;
 
-                if (avoidDuplicates.Add(storageEnvironment) == false)
-                    continue; // already seen
+                if (force == false && envSyncReq.Value.IsRequired == false &&
+                   env.Journal.Applicator.TotalWrittenButUnsyncedBytes < env.Options.JournalsSizeThreshold)
+                    continue;
 
-                if (storageEnvironment.Disposed)
+                var exchange = Interlocked.Exchange(ref envSyncReq.Value.IsSyncRun, 1);
```

Comment:
Should use `CompareExchange` here. 
But note that we should verify that it is not hanging there with `InSync` all the time if there is an error